ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/Jumping.hs

```diff
@@ -0,0 +1,665 @@
+{-# LANGUAGE DeriveAnyClass       #-}
+{-# LANGUAGE DeriveGeneric        #-}
+{-# LANGUAGE DerivingStrategies   #-}
+{-# LANGUAGE FlexibleContexts     #-}
+{-# LANGUAGE FlexibleInstances    #-}
+{-# LANGUAGE LambdaCase           #-}
+{-# LANGUAGE NamedFieldPuns       #-}
+{-# LANGUAGE RankNTypes           #-}
+{-# LANGUAGE ScopedTypeVariables  #-}
+{-# LANGUAGE StandaloneDeriving   #-}
+{-# LANGUAGE TupleSections        #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | ChainSync jumping (CSJ) is an optimization for the ChainSync protocol that
+-- allows nodes to sync without downloading headers from all of the honest
+-- peers. This load is undesirable as it slows down all the peers involved.
+--
+-- The idea is to download the headers of a chain from a single peer (the
+-- dynamo) and then ask periodically to the other peers (the jumpers) whether
+-- they agree with the dynamo's chain.
+--
+-- When the jumpers disagree with the dynamo, the jumper with the oldest
+-- intersection is asked to compete with the dynamo in the GDD logic. If the
+-- dynamo is disconnected, a new dynamo is elected and the objector is demoted
+-- to a jumper.
+--
+-- If the objector is disconnected, the syncing process continues with the
+-- dynamo and the remaining jumpers.
+--
+-- The main property of the algorithm is that it never
+-- downloads headers from more than two plausibly honest peers at a time (a
+-- dynamo and an objector). All other peers are either waiting their turn to
+-- compete with the dynamo, or are in agreement with it, or are disengaged
+-- (see next section).
+--
+-- The algorithm might still download headers redundantly from peers that do
+-- historical rollbacks. These rollbacks, however, constitute dishonest
+-- behavior, and CSJ does not concern itself with avoiding load to dishonest
+-- peers. Avoiding the load induced by dishonest peers on the syncing node would
+-- require additionally to disconnect peers that do historical rollbacks. This
+-- is not done by CSJ.
+--
```

Comment:
```suggestion
--
-- Overview of the state transitions
-- ---------------------------------
--
-- See 'ChainSyncJumpingState' for the implementation of the states.
--
--                 j       ╔════════╗
--             ╭────────── ║ Dynamo ║ ◀─────────╮
--             │           ╚════════╝           │f
--             ▼                  ▲             │
--     ┌────────────┐             │     k     ┌──────────┐
--     │ Disengaged │ ◀───────────│────────── │ Objector │
--     └────────────┘       ╭─────│────────── └──────────┘
--                          │     │             ▲    ▲ │
--                         g│     │e         b  │    │ │
--                          │     │       ╭─────╯   i│ │c
--                  ╭╌╌╌╌╌╌╌▼╌╌╌╌╌╌╌╌╌╌╌╌╌│╌╌╌╌╌╌╌╌╌╌│╌▼╌╌╌╮
--                  ┆ ╔═══════╗  a   ┌──────┐  d   ┌─────┐ |
--                  ┆ ║ Happy ║ ───▶ │ LFI* │ ───▶ │ FI* │ |
--                  ┆ ╚═══════╝ ◀─╮  └──────┘      └─────┘ |
--                  ┆ Jumper      ╰─────┴────────────╯h    |
--                  ╰╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╯
--
-- *: LookingForIntersection and FoundIntersection, abbreviated for this
--    drawing only; this abbreviation will not be used elsewhere.
--
-- 0. A new peer starts as the dynamo if there is no other peer or as a Happy
--    jumper otherwise. The dynamo periodically requests jumps from happy
--    jumpers who, in the ideal case, accept them.
--
-- 1. In the event that a jumper rejects a jump, it goes from Happy to LFI* (a).
--    From there starts a back-and-forth of intersection search messages until
--    the exact point of disagreement with the dynamo is found.
--
-- 2. Once the exact point of disagreement is found, and if there is no objector
--    yet, the jumper becomes the objector (b). If there is an objector, then we
--    compare the intersections of the objector and the jumper with the dynamo. If
--    the jumper's intersection is strictly older, then the jumper replaces the
--    dynamo (b+c). Otherwise, the jumper is marked as FI* (d).
--
-- 3. If the dynamo disconnects, one peer is elected as the new dynamo (e|f) and
--    all other peers revert to being happy jumpers (g+h).
--
-- 4. If the objector disconnects, and there are FI* jumpers, then the one with the
--    oldest intersection with the dynamo gets elected (i).
--
-- 5. If the dynamo rolls back to a point older than the last jump it requested, it
--    is disengaged (j) and a new dynamo is elected (e|f).
--
-- 6. If the objector agrees with the dynamo, it is disengaged (k). If there are FI*
--    jumpers, then one of them gets elected as the new objector (i).
--
```

Do you think something like this is useful? I might have missed transitions, and also I'm not sure if we're implementing exactly this, in particular for (2), but that's related to my previous review comments.

Reply:
Looks good to me! It should be useful to those approaching the implementation by first time.